tendermint/src/lite/verifier.rs

```diff
@@ -154,3 +165,137 @@ where
 
     Ok(())
 }
+
+/// Returns Ok if we can trust the passed in (untrusted) header
+/// based on the given trusted state, otherwise returns an Error.
+fn can_trust<TS, SH, VS, L, S, R>(
+    trusted_state: &TS,    // h1 in spec
+    untrusted_header: &SH, // h2 in spec
+    trust_threshold: &L,
+    trusting_period: &Duration,
+    now: &SystemTime,
+    req: &R,
+    store: &mut S,
+) -> Result<(), Error>
+where
+    TS: TrustedState<LastHeader = SH, ValidatorSet = VS>,
+    SH: SignedHeader,
+    VS: ValidatorSet,
+    L: TrustThreshold,
+    S: Store<SignedHeader = SH>,
+    R: Requester<SignedHeader = SH, ValidatorSet = VS>,
+{
+    // can we trust the still untrusted header based on the given trusted state?
+    match check_support(
+        trusted_state,
+        untrusted_header,
+        trust_threshold,
+        trusting_period,
+        now,
+    ) {
+        Ok(_) => {
+            store.add(untrusted_header)?;
+            return Ok(());
+        }
+        Err(e) => {
+            if e != Error::InsufficientVotingPower {
+                return Err(e);
+            }
+        }
+    }
+
+    // Here we can't trust the passed in untrusted header based on the known trusted state.
+    // Run bisection: try again with a pivot header whose height is in the
+    // middle of the trusted height and the desired height (h2.height).
+
+    let trusted_header = trusted_state.last_header();
+    let trusted_height: u64 = trusted_header.header().height().value();
+    let untrusted_height: u64 = untrusted_header.header().height().value();
```

Comment:
Because of the `: u64` we could also write `trusted_header.header().height().into()`. Not sure that's better. We can leave away the `: u64` here (in the case we keep .value()).

Reply:
Hmm I kind of assumed having the `: u64` would eliminate the need to have an extra method invocation and that it could happen automatically if the height implements Into u64 or something

Reply:
You'd need something like:

```rust
let Height(untrusted_height) = ...
```

or:

```rust
let untrusted_height: u64 = untrusted_header...height().into()
```

or:

```rust
let untrusted_height = untrusted_header...height().0
```